### ISSUE TITLE
"pointer being freed was not allocated" on shutdown

### ISSUE THREAD

**Issue by benjamingeer** [dhlab-basel/Sipi#93]
This happens on the `develop` branch. I've made some attempts to fix it on the `non_image_file_support` branch, without success.

To reproduce, start Sipi like this:

```
$ local/bin/sipi -config config/sipi.no-knora-test-config.lua
```

Make a few requests to [http://127.0.0.1:1024/knora/Leaves.jpg/full/full/0/default.jpg](http://127.0.0.1:1024/knora/Leaves.jpg/full/full/0/default.jpg) in the browser. Then type Control-C:

```
Jan  4 20:07:25  Sipi[68525] <Info>: Server shutting down
sipi(68525,0x700000081000) malloc: *** error for object 0x7faf53f00040: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```

I tried running Sipi in `lldb` and setting a breakpoint in `malloc_error_break`, but then the debugger doesn't seem to stop at the breakpoint.

I also tried running Sipi in [Valgrind](http://valgrind.org/), which can be installed with `brew install valgrind`. I ran it like this:

```
valgrind --leak-check=yes local/bin/sipi -config config/sipi.no-knora-test-config.lua
```

Then for some reason Control-C doesn't work, so I stopped Sipi using `kill -SIGBUS`. Valgrind then says:

```
==69096==  Non-existent physical address at address 0x700002314000
==69096==    at 0x1011152A2: poll (in /usr/lib/system/libsystem_kernel.dylib)
==69096==    by 0x100090641: Sipi::SipiHttpServer::run() (SipiHttpServer.cpp:1199)
==69096==    by 0x100005F1C: main (sipi.cpp:425)
```

It also reports a lot of potential memory leaks.

On the `non_image_file_support` branch I tried to replace some raw pointers with references, and some arrays with `std::vector`, but this didn't help.

**Comment by lrosenth** [comment]
Hi Ben,

This must be a new error... I'll have a look at it tonight....

Von meinem iPad gesendet

Am 04.01.2017 um 20:25 schrieb Benjamin Geer <notifications@github.com<mailto:notifications@github.com>>:


Assigned #93<https://github.com/dhlab-basel/Sipi/issues/93> to @lrosenth<https://github.com/lrosenth>.

**Comment by lrosenth** [comment]
Hi Ben,

i can’t reproduce the problem. Everything runs fine on my machine… (OS X). I would be carefull in replacing raw pointers with something else – it is a reason why I used in certain cases raw pointers! Also, <vector> may cause problems. Let’s have a look together at the code!

Best Regards, see You soon!

	Lukas
--  

Prof. Dr. Lukas Rosenthaler
Head of "Digital Humanities Lab",
University of Basel
Bernoullistrasse 32
CH-4056 Basel, Switzerland

Tel: + 41 61 267 0490
Fax: +41 61 267 3855
http://www.dhlab.unibas.ch

**Comment by benjamingeer** [comment]
The error doesn't always happen for me, sometimes I have to try several times before it happens. Yes, let's look at it tomorrow.

>  it is a reason why I used in certain cases raw pointers

For example, in `LuaServer.h`, I didn't see any benefit to using a pointer and an array here:

```
std::vector<LuaValstruct>  executeLuafunction(const std::string *funcname, int n, LuaValstruct *lv);
```

So I changed it to this (on the `non_image_file_support` branch):

```
std::vector<LuaValstruct> executeLuafunction(const std::string& funcname, std::vector<LuaValstruct>& lvs);
```

That method is only called from two places, both in `SipiHttpServer.cpp` (and I combined them into one, since the code was just duplicated).
